[REVIEW] Make many more models picklable/cloneable/scoreable

JohnZed opened the pull request:
Fixes Issue #1531 , Issue #1530, and gets a good stab at issue #1515 

I tried dask hyperopt with a number of these models and found consistent issues:
 (1) We need to support get_params/set_params reliably to handle hyperopt
 (2) Models need to be picklable both after fitting and when they have not yet been fit (common issue)
 (3) Models need a score function

I expanded the pickle tests to automate checking item (2). I think get_params/set_params tests need more thought since it would be easy to create a nominal test that doesn't actually test thoroughly enough, so I'm holding off on that until a later change.

dantegd commented on python/cuml/solvers/qn.pyx:
```diff
@@ -418,9 +418,42 @@ class QN(Base):
         return cudf.Series(preds)
 
+    def score(self, X, y):
+        from cuml.metrics import accuracy_score
```
wouldn't this import be better outside of the method/class alongside the regular imports?

dantegd commented on python/cuml/solvers/qn.pyx:
```diff
@@ -418,9 +418,42 @@ class QN(Base):
+            if self.fit_intercept:
                 return self.coef_[-1]
+            else:
+                return rmm.to_device(np.zeros(1))
```
```suggestion
                return rmm.to_device(zeros(1))
```

Minor, but it's better to use the device `zeros` function https://github.com/rapidsai/cuml/blob/47606a516b52df369d14f8792b92919e1e2b59be/python/cuml/utils/numba_utils.py#L128

dantegd commented on python/cuml/test/test_pickle.py:
```diff
@@ -29,14 +29,18 @@
     "LinearRegression": lambda: cuml.LinearRegression(),
+    "LogisticRegression": lambda: cuml.LogisticRegression(),
+    "LogisticRegressionNoInt": lambda: cuml.LogisticRegression(
+        fit_intercept=False),
```
shouldn't a no intercept method be also added to the other linear models for consistency and future classes?

JohnZed replied:
LogisticRegression has slightly more complex intercept_ handling logic than other classes (pulling it out of qn which has its own special intercept support whereas for some other classes it's just a float), so I only worried about it here. Maybe it's still worth doing everywhere though... let me give it a little thought.

JohnZed replied:
I'll add fit_intercept checking for all regression models - thanks for pointing that out.

dantegd commented on python/cuml/test/test_pickle.py:
```diff
@@ -303,6 +299,22 @@ def assert_model(pickled_model, X_test):
+                         all_models.keys())
+def test_unfit_pickle(model_name):
+    # Any model xfailed in this test cannot be used for hyperparameter sweeps
+    # with dask or sklearn
```
This should be in the docstring of their classes (not sure if we have a note anywhere else regarding PCA and UMAP on this, also we should probably open github issues (if there exist none) and put them in the hyperparam board

JohnZed replied:
Yes, good point. Filed as Issue #1538

JohnZed replied:
Ok, this was pretty easy to fix rather than commenting. Thanks for pointing it out. Fixed in the updates.

dantegd commented on python/cuml/test/test_random_forest.py:
```diff
@@ -171,6 +171,9 @@ def test_rf_classification_default(datatype, column_info, nrows):
     cu_acc = accuracy_score(y_test, cu_preds)
 
+    score_acc = cuml_model.score(X_test, y_test)
+    assert cu_acc == pytest.approx(score_acc)
```
why were this not in the tests before?

JohnZed replied:
There was no test for score and unsurprisingly score was broken.
The broken score function prevents hyperopt from working with RF, so I fixed it and expanded the test here.

dantegd commented on python/cuml/test/test_random_forest.py:
```diff
@@ -210,6 +213,11 @@ def test_rf_regression_default(datatype, column_info, nrows):
+    # score function should be equivalent
+    score_mse = cuml_model.score(X_test, y_test)
+    manual_mse = ((fil_preds - y_test)**2).mean()
+    assert manual_mse == pytest.approx(score_mse)
```
Same question as above, can't remember why this checks are new in this PR

JohnZed replied:
As above.

dantegd commented on python/cuml/linear_model/logistic_regression.pyx:
```diff
@@ -26,10 +26,12 @@ from cuml.common.base import Base
 
 supported_penalties = ['l1', 'l2', 'none', 'elasticnet']
 
-supported_solvers = ['qn', 'lbfgs', 'owl']
```
Small question: shouldn't we deprecate `lbfgs` and `owl` for 0.12 to not just break in the (probably small chance of ) case of someone currently using? It probably won't affect many users, but this could be a good chance to use deprecation practices

JohnZed replied:
Interesting... my contention is that this has always been a bug. Choosing LBFGS or OWL as the solver has no effect in logistic regression (:man_facepalming:) so anyone using them has been getting silently ripped off in the past.

dantegd replied:
yeah I think this one is fine since owl in particular is there for L1 so it is automatically selected, this design for the supported solvers was just while things were getting sorted out so dropping it probably will have no consequences

dantegd commented on python/cuml/metrics/base.py:
```diff
@@ -32,8 +32,8 @@ def score(self, X, y, **kwargs):
-        y : [cudf.Series]
-            True values for predict(X)
+        y : [cudf.Series, device array, or numpy array]
+            Ground truth values for predict(X)
```
Maybe we could change this docstring to be the normal format, or maybe we should open an issue to go around our docstrings to check uniformity? What do you think?

JohnZed replied:
I'll open an issue. Here, there were just too many docstring changes to roll into this change, so I kept it locally consistent.

JohnZed commented:
Thanks, @dantegd ! Many helpful comments. I believe I've addressed them.

cjnolet commented on python/cuml/test/test_pickle.py:
```diff
@@ -303,6 +299,22 @@ def assert_model(pickled_model, X_test):
+       model_name in ["PCA", "UMAP"]:
+        pytest.xfail()
+
+    # Pickling should work even if fit has not been called
```
I'm trying to understand the bug in UMAP/PCA picking. Is this where they are experiencing problems?

JohnZed replied:
The issue was specifically for not-yet-fit umap/pca objects. Getstate and setstate would reference fields that hadn't been created yet. I just ensure those are set to None in the initializer.

JohnZed commented:
I added support for cloning randomforest models and added a quick test for clone (for all models), which is _slightly_ different from pickle in its needs. (This test was failing for both RF types but now passes.)

JohnZed commented:
@dantegd to re-review changes

cjnolet commented:
#1547 should have fixed UMAP's pickling problem.

dantegd commented on python/cuml/ensemble/randomforestclassifier.pyx:
```diff
@@ -302,7 +303,9 @@ class RandomForestClassifier(Base):
-        self.model_pbuf_bytes = self._get_model_info()
+        if self.n_cols:
+            # only if model has been fit previously
+            self.model_pbuf_bytes = self._get_model_info()
```
Not for this PR, but I remember you raised something about renaming this function? `_get_model_info()` seems like not the best name for this

JohnZed replied:
yup, should rename in future RF change.

dantegd commented on python/cuml/linear_model/logistic_regression.pyx:
```diff
@@ -26,10 +26,13 @@ from cuml.solvers import QN
 from cuml.common.base import Base
 from cuml.metrics.accuracy import accuracy_score
 from cuml.utils import input_to_dev_array
+from cuml.utils.cupy_utils import checked_cupy_unique
```
the error of CI comes from reintroducing this `import` since this function was deleted by #1532

JohnZed replied:
Ah, bad merge, thanks for catching!

JohnZed commented on cpp/src/randomforest/randomforest.cu:
```diff
@@ -203,7 +203,7 @@ void validity_check(const RF_params rf_params) {
          "rows_sample value %f outside permitted (0, 1] range",
          rf_params.rows_sample);
-  DecisionTree::validity_check(rf_params.tree_params);
+  // DecisionTree::validity_check(rf_params.tree_params);
```
Ah, I think we can uncomment that now. It was left off from a debugging exercise. Will fix.

dantegd commented on CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@
 - PR #1528: Re-enabling KNN using dynamic library loading for UCX in communicator
 - PR #1545: Add conda environment version updates to ci script
 - PR #1541: Updates for libcudf++ Python refactor
+- PR #1531: Improve pickling and scoring suppport for many models to support hyperopt
```
PR number is wrong, that's why the changelog CI is failing

JohnZed replied:
Fixed, thanks!

cjnolet commented:
rerun tests

cjnolet commented on python/cuml/manifold/umap.pyx:
```diff
@@ -313,6 +316,8 @@ class UMAP(Base):
         self.umap_params = <size_t> umap_params
 
         self.callback = callback  # prevent callback destruction
+        self.X_m = None
```
I'm still seeing this in the CI log so I think UMAP is still having pickling issues:
```
self = <class 'cudf.core.dataframe.DataFrame'>, data = None, index = None

columns = None, nan_as_null = False



    @classmethod

    def from_gpu_matrix(

        self, data, index=None, columns=None, nan_as_null=False

    ):

        """Convert from a numba gpu ndarray.

    

        Parameters

        ----------

        data : numba gpu ndarray

        index : str

            The name of the index column in *data*.

            If None, the default index is used.

        columns : list of str

            List of column names to include.

    

        Returns

        -------

        DataFrame

        """

>       if data.ndim != 2:

E       AttributeError: 'NoneType' object has no attribute 'ndim'
```

This is happening because I'm using `hasattr` in the serialization functions and you are setting them to None in the constructor, causing it to call `None.from_gpu_matrix()` in the deserialization function. 

I think we should either 1) not set the values to None in the constructor, or 2) check for `hasattr() and is not None`. Otherwise, I think this PR is ready to go.

JohnZed commented:
rerun tests

JohnZed commented:
rerun tests

dantegd commented:
rerun tests
